Shell/Formatter.cpp

```diff
@@ -0,0 +1,648 @@
+    current_builder().append(" &");
+}
+
+void Formatter::visit(const AST::BarewordLiteral* node)
+{
+    test_and_update_output_cursor(node);
+    current_builder().append(node->text());
+}
+
+void Formatter::visit(const AST::CastToCommand* node)
+{
+    test_and_update_output_cursor(node);
+    if (m_options.explicit_parentheses)
+        current_builder().append('(');
+
+    TemporaryChange<const AST::Node*> parent { m_parent_node, node };
+    NodeVisitor::visit(node);
+
+    if (m_options.explicit_parentheses)
+        current_builder().append(')');
+}
+
+void Formatter::visit(const AST::CastToList* node)
+{
+    test_and_update_output_cursor(node);
+    current_builder().append('(');
+
+    TemporaryChange<const AST::Node*> parent { m_parent_node, node };
+    NodeVisitor::visit(node);
+
+    current_builder().append(')');
+}
+
+void Formatter::visit(const AST::CloseFdRedirection* node)
+{
+    test_and_update_output_cursor(node);
+    TemporaryChange<const AST::Node*> parent { m_parent_node, node };
+
+    current_builder().appendf(" %d>&-", node->fd());
+}
+
+void Formatter::visit(const AST::CommandLiteral*)
+{
+    ASSERT_NOT_REACHED();
+}
+
+void Formatter::visit(const AST::Comment* node)
+{
+    test_and_update_output_cursor(node);
+    current_builder().append("#");
+    current_builder().append(node->text());
+}
+
+void Formatter::visit(const AST::DynamicEvaluate* node)
+{
+    test_and_update_output_cursor(node);
+    current_builder().append('$');
+    TemporaryChange<const AST::Node*> parent { m_parent_node, node };
+    NodeVisitor::visit(node);
+}
+
+void Formatter::visit(const AST::DoubleQuotedString* node)
+{
+    test_and_update_output_cursor(node);
+    current_builder().append("\"");
+
+    TemporaryChange quotes { m_options.in_double_quotes, true };
+    TemporaryChange<const AST::Node*> parent { m_parent_node, node };
+
+    NodeVisitor::visit(node);
+
+    current_builder().append("\"");
+}
+
+void Formatter::visit(const AST::Fd2FdRedirection* node)
+{
+    test_and_update_output_cursor(node);
+    TemporaryChange<const AST::Node*> parent { m_parent_node, node };
+
+    current_builder().appendf(" %d>&%d", node->source_fd(), node->dest_fd());
+    if (m_hit_node == node)
+        ++m_output_cursor;
+}
+
+void Formatter::visit(const AST::FunctionDeclaration* node)
+{
+    test_and_update_output_cursor(node);
+    current_builder().append(node->name().name);
+    current_builder().append('(');
+    TemporaryChange<const AST::Node*> parent { m_parent_node, node };
+
+    auto first = true;
+    for (auto& arg : node->arguments()) {
+        if (!first)
+            current_builder().append(' ');
+        first = false;
+        current_builder().append(arg.name);
+    }
+
+    current_builder().append(") {");
+
+    {
+        TemporaryChange indent { m_current_indent, m_current_indent + 1 };
+        insert_separator();
+
+        if (node->block())
+            node->block()->visit(*this);
+    }
```

Comment:
Do we need this block?

Reply:
We do, `TemporaryChange` reverts the indentation when this block ends.

Reply:
Thanks for raising this issue.
I abstracted this stuff away in `in_new_block()` and `with_added_indent()`.